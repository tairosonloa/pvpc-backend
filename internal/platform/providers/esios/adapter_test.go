package esios

import (
	"context"
	"net/http"
	"net/http/httptest"
	"os"
	"testing"
	"time"

	"pvpc-backend/internal/domain"
	"pvpc-backend/internal/domain/errors"

	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/stretchr/testify/require"
)

const MOCK_TOKEN = "FAKE_TOKEN"

func Test_FetchPVPCPrices_Success(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		require.Equal(t, http.MethodGet, r.Method)
		require.Equal(t, pvpcPricesEndpoint, r.URL.Path)
		require.Equal(t, "application/json", r.Header.Get("accept"))
		require.Equal(t, MOCK_TOKEN, r.Header.Get("x-api-key"))
		require.Equal(t, "end_date=2023-09-08T23%3A59%3A59&geo_ids%5B%5D=1234&geo_ids%5B%5D=5678&start_date=2023-09-08T00%3A00%3A00", r.URL.RawQuery)

		res, err := os.ReadFile("./mocks/fetch_pvpc_response.json")
		require.NoError(t, err)

		w.WriteHeader(http.StatusOK)
		w.Write(res)
	}))
	defer server.Close()

	zone1, err := domain.NewZone(domain.ZoneDto{ID: "FOO", ExternalID: "1234", Name: "Foo Zone"})
	require.NoError(t, err)
	zone2, err := domain.NewZone(domain.ZoneDto{ID: "BAR", ExternalID: "5678", Name: "Bar Zone"})
	require.NoError(t, err)

	date, err := time.Parse("2006-01-02T15:04:05Z", "2023-09-08T17:54:36Z")
	require.NoError(t, err)

	adapter := NewEsiosAPI(server.URL, MOCK_TOKEN)
	prices, err := adapter.FetchPVPCPrices(context.Background(), []domain.Zone{zone1, zone2}, date)
	require.NoError(t, err)
	require.Len(t, prices, 2)
	snaps.MatchSnapshot(t, prices[0].Serialize(), prices[1].Serialize())
}

func Test_FetchPVPCPrices_Error(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		require.Equal(t, http.MethodGet, r.Method)
		require.Equal(t, pvpcPricesEndpoint, r.URL.Path)
		require.Equal(t, "application/json", r.Header.Get("Accept"))
		require.Equal(t, "end_date=2023-09-08T23%3A59%3A59&geo_ids%5B%5D=1234&start_date=2023-09-08T00%3A00%3A00", r.URL.RawQuery)

		w.WriteHeader(http.StatusTooManyRequests)
		w.Write([]byte("error"))
	}))
	defer server.Close()

	zone, err := domain.NewZone(domain.ZoneDto{ID: "ZON", ExternalID: "1234", Name: "Zone Name"})
	require.NoError(t, err)

	date, err := time.Parse("2006-01-02T15:04:05Z", "2023-09-08T17:54:36Z")
	require.NoError(t, err)

	adapter := NewEsiosAPI(server.URL, MOCK_TOKEN)
	prices, err := adapter.FetchPVPCPrices(context.Background(), []domain.Zone{zone}, date)
	require.Error(t, err)
	require.Equal(t, errors.ProviderError, errors.Code(err))
	require.Len(t, prices, 0)
}
