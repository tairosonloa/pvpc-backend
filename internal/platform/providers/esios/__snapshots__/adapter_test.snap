
[Test_FetchPVPCPrices_Success - 1]
domain.PricesDto{
    ID:     "FOO-2023-09-08",
    Date:   "2023-09-08",
    Zone:   domain.ZoneDto{ID:"FOO", ExternalID:"1234", Name:"Foo Zone"},
    Values: {
        {Datetime:"2023-09-08T00:00:00+02:00", Value:151.9600067138672},
        {Datetime:"2023-09-08T01:00:00+02:00", Value:146.2100067138672},
        {Datetime:"2023-09-08T02:00:00+02:00", Value:144.6999969482422},
        {Datetime:"2023-09-08T03:00:00+02:00", Value:144.24000549316406},
        {Datetime:"2023-09-08T04:00:00+02:00", Value:143.22999572753906},
        {Datetime:"2023-09-08T05:00:00+02:00", Value:145.5500030517578},
        {Datetime:"2023-09-08T06:00:00+02:00", Value:147.0399932861328},
        {Datetime:"2023-09-08T07:00:00+02:00", Value:145.02999877929688},
        {Datetime:"2023-09-08T08:00:00+02:00", Value:163.7899932861328},
        {Datetime:"2023-09-08T09:00:00+02:00", Value:162.89999389648438},
        {Datetime:"2023-09-08T10:00:00+02:00", Value:208.94000244140625},
        {Datetime:"2023-09-08T11:00:00+02:00", Value:186.25},
        {Datetime:"2023-09-08T12:00:00+02:00", Value:185.08999633789062},
        {Datetime:"2023-09-08T13:00:00+02:00", Value:184.44000244140625},
        {Datetime:"2023-09-08T14:00:00+02:00", Value:137.7100067138672},
        {Datetime:"2023-09-08T15:00:00+02:00", Value:136.75},
        {Datetime:"2023-09-08T16:00:00+02:00", Value:136.27999877929688},
        {Datetime:"2023-09-08T17:00:00+02:00", Value:147.47000122070312},
        {Datetime:"2023-09-08T18:00:00+02:00", Value:206.8800048828125},
        {Datetime:"2023-09-08T19:00:00+02:00", Value:216.52000427246094},
        {Datetime:"2023-09-08T20:00:00+02:00", Value:228.11000061035156},
        {Datetime:"2023-09-08T21:00:00+02:00", Value:234.52000427246094},
        {Datetime:"2023-09-08T22:00:00+02:00", Value:188.72000122070312},
        {Datetime:"2023-09-08T23:00:00+02:00", Value:178.6300048828125},
    },
}
domain.PricesDto{
    ID:     "BAR-2023-09-08",
    Date:   "2023-09-08",
    Zone:   domain.ZoneDto{ID:"BAR", ExternalID:"5678", Name:"Bar Zone"},
    Values: {
        {Datetime:"2023-09-08T00:00:00+02:00", Value:151.9600067138672},
        {Datetime:"2023-09-08T01:00:00+02:00", Value:146.2100067138672},
        {Datetime:"2023-09-08T02:00:00+02:00", Value:144.6999969482422},
        {Datetime:"2023-09-08T03:00:00+02:00", Value:144.24000549316406},
        {Datetime:"2023-09-08T04:00:00+02:00", Value:143.22999572753906},
        {Datetime:"2023-09-08T05:00:00+02:00", Value:145.5500030517578},
        {Datetime:"2023-09-08T06:00:00+02:00", Value:147.0399932861328},
        {Datetime:"2023-09-08T07:00:00+02:00", Value:145.02999877929688},
        {Datetime:"2023-09-08T08:00:00+02:00", Value:163.7899932861328},
        {Datetime:"2023-09-08T09:00:00+02:00", Value:162.89999389648438},
        {Datetime:"2023-09-08T10:00:00+02:00", Value:208.94000244140625},
        {Datetime:"2023-09-08T11:00:00+02:00", Value:186.25},
        {Datetime:"2023-09-08T12:00:00+02:00", Value:185.08999633789062},
        {Datetime:"2023-09-08T13:00:00+02:00", Value:184.44000244140625},
        {Datetime:"2023-09-08T14:00:00+02:00", Value:137.7100067138672},
        {Datetime:"2023-09-08T15:00:00+02:00", Value:136.75},
        {Datetime:"2023-09-08T16:00:00+02:00", Value:136.27999877929688},
        {Datetime:"2023-09-08T17:00:00+02:00", Value:147.47000122070312},
        {Datetime:"2023-09-08T18:00:00+02:00", Value:206.8800048828125},
        {Datetime:"2023-09-08T19:00:00+02:00", Value:216.52000427246094},
        {Datetime:"2023-09-08T20:00:00+02:00", Value:228.11000061035156},
        {Datetime:"2023-09-08T21:00:00+02:00", Value:234.52000427246094},
        {Datetime:"2023-09-08T22:00:00+02:00", Value:188.72000122070312},
        {Datetime:"2023-09-08T23:00:00+02:00", Value:178.6300048828125},
    },
}
---
